ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/sink/DataGroup.scala

```
@@ -44,6 +61,7 @@ private[sink] class DataGroup(val name: String, brokerProps: String, topicNames:
   override def close(): Unit =
     if (closed.compareAndSet(false, true)) {
       producer.close()
```

it would be better to swallow all exceptions temporarily before closing all resources. Those exceptions can squashed to single one before throwing.

---

Pardon me, not sure I'm catch your point. Can you give a example ?

---

```scala
Releasable.close(producer)
Releasable.close(rowCounter)
```
this makes sure all resources are closed even if exception happens. For another, the thrown exceptions can be amassed and then merged to one exception. For example:
```scala
      var e: Throwable = null
      Releasable.close(producer, (exception: Throwable) => if (e == null) e = exception else e.addSuppressed(exception))
      if (e != null) throw e
```

the above code may be not necessary if you feel it is fine to just swallow exception in closing objects.

---

Got it, thanks @chia7712 !